test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py

```diff
@@ -0,0 +1,506 @@
+from inspect import isclass
+from typing import List, Iterable, TypeVar, Type, NewType
+from typing import Union
+from typing_inspect import get_origin
+
+# SSZ integers
+# -----------------------------
+
+
+class uint(int):
+    byte_len = 0
+
+    def __new__(cls, value, *args, **kwargs):
+        if value < 0:
+            raise ValueError("unsigned types must not be negative")
+        return super().__new__(cls, value)
+
+
+class uint8(uint):
+    byte_len = 1
+
+    def __new__(cls, value, *args, **kwargs):
+        if value.bit_length() > 8:
+            raise ValueError("value out of bounds for uint8")
+        return super().__new__(cls, value)
+
+
+# Alias for uint8
+byte = NewType('byte', uint8)
+
+
+class uint16(uint):
+    byte_len = 2
+
+    def __new__(cls, value, *args, **kwargs):
+        if value.bit_length() > 16:
+            raise ValueError("value out of bounds for uint16")
+        return super().__new__(cls, value)
+
+
+class uint32(uint):
+    byte_len = 4
+
+    def __new__(cls, value, *args, **kwargs):
+        if value.bit_length() > 32:
+            raise ValueError("value out of bounds for uint16")
+        return super().__new__(cls, value)
+
+
+# We simply default to uint64. But do give it a name, for readability
+uint64 = NewType('uint64', int)
```

Comment:
Seems prudent to create a child class with the `bit_length` bounds check like the other `uint`s

Reply:
Is this a sacrifice to make it easier to deal with integer types in most cases?

Reply:
Previously we had `NewType` for every integer type. But those are not classes, i.e. we can't infer the type from the value, we have to get it from the annotations, and act as if the NewType reference is a type class. Then it changed to this, so we can infer a type from a value easily. We don't really use the non-uint64 types, and if we do, we may want to check their bounds more aggressively, so hence the choice to wrap them. And uint64 as default, to not box every single integer (performance + readability)

Reply:
We could make the uint types more strict in underflow/overflow checks, but that was originally an idea pushed by Dankrad, and I'd rather have him implement it his way, or pull in a library that does exactly this (fixed-size over/under-flow checked unsigned ints). This minimal form will do for now.

Reply:
Discussed in call, uint types can be extended in a later PR when necessary.